### PR TITLE
`Experiment.start` Returns Records

### PR DESCRIPTION
**Pull request by MattToast**
Basic implementation of a `Record` class to keep track of launched jobs. Convert `Experiment.start` return `Record`s and all other methods to accept records.

### CODE REVIEW

File: smartsim/experiment.py [hunk 1]
```diff
@@ -158,7 +157,7 @@ def __init__(self, name: str, exp_path: str | None = None):
         experiment
         """
 
-    def start(self, *jobs: Job | t.Sequence[Job]) -> tuple[LaunchedJobID, ...]:
+    def start(self, *jobs: Job | t.Sequence[Job]) -> tuple[Record, ...]:
```
Comment (amandarichardsonn) [inline smartsim/experiment.py:160]: Do you need to update the :return: in the start docstring?

File: smartsim/experiment.py [hunk 2]
```diff
@@ -158,7 +157,7 @@ def __init__(self, name: str, exp_path: str | None = None):
...
         """Execute a collection of `Job` instances.
```
Comment (amandarichardsonn) [inline smartsim/experiment.py:161]: It looks like there is a :raises TypeError: but I do not see a TypeError raised in this start?

File: smartsim/experiment.py [hunk 3]
```diff
@@ -233,9 +231,7 @@ def execute_dispatch(generator: Generator, job: Job, idx: int) -> LaunchedJobID:
             execute_dispatch(generator, job, idx) for idx, job in enumerate(jobs, 1)
         )
 
-    def get_status(
-        self, *ids: LaunchedJobID
-    ) -> tuple[JobStatus | InvalidJobStatus, ...]:
+    def get_status(self, *records: Record) -> tuple[JobStatus | InvalidJobStatus, ...]:
         """Get the status of jobs launched through the `Experiment` from their
         launched job id returned when calling `Experiment.start`.
```
Comment (amandarichardsonn) [inline smartsim/experiment.py]: docstring nit: job id -> Record?

File: smartsim/launchable/job.py [hunk 4]
```diff
@@ -158,3 +159,35 @@ def __str__(self) -> str:  # pragma: no cover
         string = f"SmartSim Entity: {self.entity}\n"
         string += f"Launch Settings: {self.launch_settings}"
         return string
+
+
+@t.final
+class Record:
+    """A record a job that was launched and launch ID assigned to the
```
Comment (amandarichardsonn) [inline smartsim/launchable/job.py]: doc string nit `A record a job that was launched and launch ID assigned to the launching` -> `A Record object to track a launched job along with its assigned launch ID`?

File: smartsim/launchable/job.py [hunk 5]
```diff
@@ -158,3 +159,35 @@ def __str__(self) -> str:  # pragma: no cover
...
+    launching.
+    """
+
+    def __init__(self, launch_id: LaunchedJobID, job: Job) -> None:
+        """Initialize a new record of a launched job
+
+        :param launch_id: A unique identifier for the launch of the job.
+        :param job: The job that was launched.
+        """
+        self._id = launch_id
```
Comment (amandarichardsonn) [inline smartsim/launchable/job.py:177]: Why do you deep copying the Job but not the launch_id here?

File: smartsim/launchable/job.py [hunk 6]
```diff
@@ -158,3 +159,35 @@ def __str__(self) -> str:  # pragma: no cover
...
+    def launched_id(self) -> LaunchedJobID:
+        """The unique identifier for the launched job.
+
+        :returns: A unique identifier for the launched job.
+        """
+        return self._id
+
+    @property
+    def job(self) -> Job:
+        """A deep copy of the job that was launched.
```
Comment (amandarichardsonn) [inline smartsim/launchable/job.py:190]: would adding `def __str__(self)` to the Record class be beneficial? For when a user logs the records or prints

File: smartsim/experiment.py [hunk 7]
```diff
@@ -445,20 +439,20 @@ def summary(self, style: str = "github") -> str:
...
             respective of the order of the calling arguments.
         """
-        if ids:
-            if not all(isinstance(id, str) for id in ids):
-                raise TypeError("ids argument was not of type LaunchedJobID")
-        else:
-            raise ValueError("No job ids provided")
+        if not records:
+            raise ValueError("No records provided")
+        if not all(isinstance(reocrd, Record) for reocrd in records):
```
Comment (amandarichardsonn) [inline smartsim/experiment.py]: it looks like you have `if not all(isinstance(reocrd, Record)` where record is misspelled

**Review by amandarichardsonn — CHANGES_REQUESTED** [review]
Some small suggestions!

File: smartsim/experiment.py (hunk 1 above)
Comment (MattToast) [reply smartsim/experiment.py:160]: d'oh! 🤦 I knew I let  a few of these slip through the cracks... yeah! that should be updated!!

File: smartsim/experiment.py (hunk 2 above)
Comment (MattToast) [reply smartsim/experiment.py:161]: Whoops! I think that check got moved into `unpack` and we forgot to update the docstrings. I can move it there rq!

File: smartsim/launchable/job.py (hunk 4 above)
Comment (MattToast) [reply smartsim/launchable/job.py]: Every time I read my own writing, I become more and more convinced that I'm not actually fluent in English, hahaha. Updated to make it coherent!

File: smartsim/launchable/job.py (hunk 5 above)
Comment (MattToast) [reply smartsim/launchable/job.py:177]: In theory, the `launch_id` (at least for now), is just a string that we ask the type checker to treat like a different type, and strings are immutable in python, so a "deepcopy of" and a "reference to" a string are functionally equivalent.

There is actually some special logic in [`copy.deepcopy`](https://github.com/python/cpython/blob/3.13/Lib/copy.py#L119-L137) so that, by default, [it will just return the original reference to a string](https://github.com/python/cpython/blob/3.13/Lib/copy.py#L173-L183) if its encounter without change.

TL;DR I was lazy and didn't want to spell out "deepcopy". More than willing to add it though if we think that `LaunchedJobID` might become a more substantive type in future!

File: smartsim/launchable/job.py (hunk 6 above)
Comment (MattToast) [reply smartsim/launchable/job.py:190]: Probably not a bad idea! I'll add one!

File: smartsim/experiment.py (hunk 7 above)
Comment (MattToast) [reply smartsim/experiment.py]: This is what I get for relying too much on intellisense lol

File: smartsim/launchable/job.py (hunk 6 above)
Comment (MattToast) [reply smartsim/launchable/job.py:190]: Added in [2e367a0](https://github.com/CrayLabs/SmartSim/pull/744/commits/2e367a0defd708d7a71e8c233901d7e9686d0fc4)

File: smartsim/experiment.py [hunk 8]
```diff
@@ -233,15 +230,13 @@ def execute_dispatch(generator: Generator, job: Job, idx: int) -> LaunchedJobID:
...
-        """Get the status of jobs launched through the `Experiment` from their
-        launched job id returned when calling `Experiment.start`.
+    def get_status(self, *records: Record) -> tuple[JobStatus | InvalidJobStatus, ...]:
+        """Get the status of jobs launched through the `Experiment` from the
+        record returned when calling `Experiment.start`.
 
-        The `Experiment` will map the launched ID back to the launcher that
-        started the job and request a status update. The order of the returned
-        statuses exactly matches the order of the launched job ids.
+        The `Experiment` will map the launched id of the record back to the
```
Comment (ankona) [inline smartsim/experiment.py]: the ID isn't launched.  could just say `will map the record id to the ...`

File: smartsim/experiment.py [hunk 9]
```diff
@@ -233,15 +230,13 @@ def execute_dispatch(generator: Generator, job: Job, idx: int) -> LaunchedJobID:
...
+        launcher that started the job and request a status update. The order of
+        the returned statuses exactly matches the order of the records.
```
Comment (ankona) [inline smartsim/experiment.py]: can omit exactly and just say `matches`

File: smartsim/experiment.py [hunk 10]
```diff
@@ -445,20 +438,20 @@ def summary(self, style: str = "github") -> str:
...
         :returns: A tuple of job statuses upon cancellation with order
             respective of the order of the calling arguments.
         """
-        if ids:
-            if not all(isinstance(id, str) for id in ids):
-                raise TypeError("ids argument was not of type LaunchedJobID")
-        else:
-            raise ValueError("No job ids provided")
+        if not records:
+            raise ValueError("No records provided")
```
Comment (ankona) [inline smartsim/experiment.py]: consider making the log more descriptive about what was attempted "No records provided. No jobs will be stopped."

File: smartsim/launchable/job.py [hunk 11]
```diff
@@ -158,3 +160,44 @@ def __str__(self) -> str:  # pragma: no cover
         string = f"SmartSim Entity: {self.entity}\n"
         string += f"Launch Settings: {self.launch_settings}"
         return string
+
+
+@t.final
+class Record:
+    """A Record object to track a launched job along with its assigned
```
Comment (ankona) [inline smartsim/launchable/job.py]: does it really "track" the job or would something like `A unique identifier for a launched job` express what it does?

File: smartsim/launchable/job.py [hunk 12]
```diff
@@ -158,3 +160,44 @@ def __str__(self) -> str:  # pragma: no cover
...
+        :returns: A deep copy of the launched job.
+        """
+        return deepcopy(self._job)
+
+    def __str__(self) -> str:
+        return textwrap.dedent(f"""\
+            Launch Record:
+                Launched Job ID:
+                    {self.launched_id}
+                Laucnehd Job:
```
Comment (ankona) [inline smartsim/launchable/job.py]: typo in laucnehd

File: smartsim/experiment.py (hunk 8 above)
Comment (MattToast) [reply smartsim/experiment.py]: Rephrased!

File: smartsim/experiment.py (hunk 9 above)
Comment (MattToast) [reply smartsim/experiment.py]: "exactly" dropped!

File: smartsim/experiment.py (hunk 10 above)
Comment (MattToast) [reply smartsim/experiment.py]: More detailed message issued!

File: smartsim/launchable/job.py (hunk 11 above)
Comment (MattToast) [reply smartsim/launchable/job.py]: Good point! I updated the docstring to be a bit more descriptive of the composition of the object!

File: smartsim/launchable/job.py (hunk 12 above)
Comment (MattToast) [reply smartsim/launchable/job.py]: 🤦
